test/compiler/inline.jl

```diff
@@ -1259,3 +1259,21 @@ end
 @test fully_eliminated() do
     return maybe_error_int(1)
 end
+
+@testset "type-based offset axes check" begin
+    a = randn(ComplexF64, 10)
+    ta = reinterpret(Float64, a)
+    tb = reinterpret(Float64, view(a, 1:2:10))
+    tc = reinterpret(Float64, reshape(view(a, 1:3:10), 2, 2, 1))
+    # Issue #44040
+    @test fully_eliminated(Base.require_one_based_indexing, Base.typesof(ta,tc))
+    @test fully_eliminated(Base.require_one_based_indexing, Base.typesof(tc,tc))
+    @test fully_eliminated(Base.require_one_based_indexing, Base.typesof(ta,tc,tb))
+    # Ranges && CartesianIndices
+    @test fully_eliminated(Base.require_one_based_indexing, Base.typesof(1:10,Base.OneTo(10),1.0:2.0,LinRange(1.0,2.0,2),1:2:10,CartesianIndices((1:2:10,1:2:10))))
+    # Remind us to call `any` in `Base.has_offset_axes` once our compiler is ready.
+    @inline _has_offset_axes(A) = @inline any(x->Int(first(x))::Int != 1, axes(A))
+    @inline _has_offset_axes(As...) = @inline any(_has_offset_axes, As)
+    a, b = zeros(2,2,2), zeros(2,2)
+    @test_broken fully_eliminated(_has_offset_axes, Base.typesof(a,a,b,b))
+end
```

Comment:
Maybe we want to put this test case in test/abstractarray.jl? We can `include("test/compiler/irutils.jl")`.